vertical-pod-autoscaler/e2e/actuation.go

```diff
@@ -39,54 +39,80 @@ var _ = ActuationSuiteE2eDescribe("Actuation", func() {
+		err := assertPodsPendingForDuration(f.ClientSet, d, 1, 2*time.Minute)
 		gomega.Expect(err).NotTo(gomega.HaveOccurred())
 
-		ginkgo.By("Setting up a VPA CRD with ridiculous request")
-		config, err := framework.LoadConfig()
+	})
+
+	ginkgo.It("never applies recommendations when update mode is Off", func() {
+		ginkgo.By("Setting up a hamster deployment")
+		d := SetupHamsterDeployment(f, "100m", "100Mi", defaultHamsterReplicas)
+		cpuRequest := getCPURequest(d.Spec.Template.Spec)
+		podList, err := GetHamsterPods(f)
 		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		podSet := MakePodSet(podList)
 
-		vpaCRD := NewVPA(f, "hamster-vpa", &metav1.LabelSelector{
-			MatchLabels: map[string]string{
-				"app": "hamster",
-			},
-		})
+		ginkgo.By("Setting up a VPA CRD in mode Off")
+		SetupVPA(f, "200m", vpa_types.UpdateModeOff)
 
-		resourceList := apiv1.ResourceList{
-			apiv1.ResourceCPU: ParseQuantityOrDie("9999"), // Request 9999 CPUs to make POD pending
-		}
+		ginkgo.By(fmt.Sprintf("Waiting for pods to be restarted, hoping it won't happen, sleep for %s", VpaEvictionTimeout.String()))
+		CheckNoPodsRestarted(f, podSet)
+		ginkgo.By("Forcefully killing one pod")
+		killPod(f, podList)
 
-		vpaCRD.Status.Recommendation = &vpa_types.RecommendedPodResources{
-			ContainerRecommendations: []vpa_types.RecommendedContainerResources{{
-				ContainerName: "hamster",
-				Target:        resourceList,
-				LowerBound:    resourceList,
-				UpperBound:    resourceList,
-			}},
+		ginkgo.By("Checking the requests were not modified")
+		updatedPodList, err := GetHamsterPods(f)
+		for _, pod := range updatedPodList.Items {
+			gomega.Expect(getCPURequest(pod.Spec)).To(gomega.Equal(cpuRequest))
 		}
+	})
 
-		vpaClientSet := vpa_clientset.NewForConfigOrDie(config)
-		vpaClient := vpaClientSet.AutoscalingV1beta1()
-		_, err = vpaClient.VerticalPodAutoscalers(ns).Create(vpaCRD)
-		gomega.Expect(err).NotTo(gomega.HaveOccurred())
-
-		ginkgo.By("Waiting for pods to be restarted and stuck pending")
-
-		err = assertPodsPendingForDuration(c, d, 1, 2*time.Minute)
+	ginkgo.It("applies recommendations only on restart when update mode is Initial", func() {
+		ginkgo.By("Setting up a hamster deployment")
+		SetupHamsterDeployment(f, "100m", "100Mi", defaultHamsterReplicas)
+		podList, err := GetHamsterPods(f)
 		gomega.Expect(err).NotTo(gomega.HaveOccurred())
-
+		podSet := MakePodSet(podList)
+
+		ginkgo.By("Setting up a VPA CRD in mode Off")
+		SetupVPA(f, "200m", vpa_types.UpdateModeInitial)
+		updatedCPURequest := ParseQuantityOrDie("200m")
+
+		ginkgo.By(fmt.Sprintf("Waiting for pods to be restarted, hoping it won't happen, sleep for %s", VpaEvictionTimeout.String()))
+		CheckNoPodsRestarted(f, podSet)
+		ginkgo.By("Forcefully killing one pod")
+		killPod(f, podList)
+
+		ginkgo.By("Checking that request was modified after forceful restart")
+		updatedPodList, err := GetHamsterPods(f)
+		foundUpdated := false
```

Comment:
I think the test would be slightly stronger if foundUpdated was an int and compared against the expected value of 1.

Reply:
Done